Split block position in SQLite3 database

ShadowNinja opened the pull request:
Old databases are automagically converted when they're loaded.

Rationale: Cramming all of the position fields into one field is just messy.  But besides that, it's also a lot harder to process, so many tools that operate on databases have to iterate over a large portion of the database, decoding and checking positions, rather than using a simple SQL statement to fetch only the blocks they need.

This also switches to R*Tree based storage.  This _may_ improve performance, but it will only be substantial with very big databases.  Small databases _will_ be slower with the R*Tree.

Real-world improvements:
1. The mapper(s) can load only the blocks that they need.  The C++ mapper currently loads whole Z-slices at a time, rather than just blocks in it's generation range, because it can't limit the selection through SQL.
2. Some map editing tools that I've written would be substantially shorter or could be done with only SQL.  Eg, a script that removed all block that were more that a certain distance from the origin could be done with `DELETE FROM blocks WHERE ABS(x) > 3 OR ABS(y) > 3 ABS(z) > 3;`.
3. Corrupted blocks can be deleted without having to write a script to convert the block position to a position hash.

I've converted a map with over 100,000 blocks in about ten seconds.

TODO:
- Update the C++ mapper.

ShadowNinja commented:
Rebased.

rubenwardy commented:
:+1:

This makes sense.

rubenwardy commented on src/database-redis.cpp:
```diff
@@ -127,18 +129,14 @@ bool Database_Redis::deleteBlock(v3s16 blockpos)
-	redisReply *reply = (redisReply *)redisCommand(ctx, "HDEL %s %s",
-		hash.c_str(), tmp.c_str());
+	redisReply *reply = static_cast<redisReply *>(redisCommand(ctx,
+		"HDEL %s %s", hash.c_str(), tmp.c_str()));
```
Not that it matters, there should be two levels of indentation from `redisReply` here.

Zeno- commented:
@rubenwardy What you said is true (yes, two adjacent string literals with only whitespace between them would be concatenated). The only difference, really, is that string literals have an implementation defined maximum length (but a conforming compiler must accommodate string literals up to at least 4095 characters in length after concatenation). I don't think std::string has any such limitation. Although this is not an issue here at all -- although I suppose hypothetically it could be.

ShadowNinja commented:
Rebased.

nerzhul commented:
This commit is very very interesting. Could you rebase and squash commits ?

PilzAdam commented:
Why do you add an id column? Do you use it anywhere?

ShadowNinja commented:
@PilzAdam No, it isn't currently used.

@nerzhul I'd like to know that this will be merged before a rebase and squash it yet again.

This fixes some client map saving issues separate from the pos split too (eg, map isn't saved to memory and then all written at once when the client shuts down, and the Client doesn't spawn a server anymore, fixing #2159)

PilzAdam commented:
@ShadowNinja Why not use x, y and z as primar keys then?

ShadowNinja commented:
@PilzAdam No particular reason.  I don't know of a good reason to do it that way though.

SQLite3 auto-generates an `INTEGER PRIMARY KEY` column by default if you don't (and lets you access it as `rowid`, `oid`, or `_rowid_`) so there isn't really a reason not to declare it ourselves.

We could use the `WITHOUT ROWID` clause to remove that, it would save space and maybe be a little faster (or slower).  It breaks things like BLOb IO, `update_hook()` and `last_insert_rowid()` though.  I'm not sure what would work best.  (http://www.sqlite.org/withoutrowid.html).

PilzAdam commented:
@ShadowNinja The best reason to make the coordinates the key is because they *are* the key.

rubenwardy commented:
I agree that it should use a composite primary key. The only reason not to use one would be if it is used as a foreign key in other tables and it would increase the database size or you want to change the key (for example userID instead of making the username a primary key).

rubenwardy commented:
Other database backends should be updated as well.

s-l-teichmann commented:
Have you considered to store the key in an bitwise interleaved format instead of splitting the position into three separate columns?

This gives you the effect of having a spatial index for the blocks (keywords: Z order, Morton Curve, linear Octree). Queries of the kind "Give me all block in the cuboid (x1, y1, y2) - (x2, y2, z2)" can be done pretty efficient.

This is one of the central keys to the performance of the MTSatellite live mapping system. Even large maps are prerendered within in minutes with very low memory consumption. Our ~4GB world is renderered incl. zoom tiles on 6 cores in about 4 minutes. Updating this map take only fractions of a seconds. We use a db cache size of 64MB. And all this at same same time with a number of players are actively using the MT server.  Yes its fast ... damn fast.

If you are interessed in the details I'm willing to elaborate more details.

Rogier-5 commented:
Hi,
I thought I'd give the new database format a try with minetestmapper.

I haven't gotten around to actually doing that yet, but I did run some conversion speed tests between sqlite and leveldb with the old and new format, and a 5GB world. I thought I'd share the results:

> Original: sqlite -> leveldb: 332.57user 162.38system 12:44.17elapsed 64%CPU
> Original: leveldb -> sqlite: 340.24user 130.84system 8:10.76elapsed 95%CPU

> New: sqlite -> leveldb: 792.14user 332.93system 34:38.71elapsed 54%CPU
> New: leveldb -> sqlite: 446.28user 247.77system 13:56.07elapsed 83%CPU

In-place database conversion took about 20 minutes.

(tests were run on an SSD)

BTW: filesys.cpp must be compiled with `#define _FILE_OFFSET_BITS 64` for the automatic conversion to work on linux (else stat() fails for database files > 4G, even if they do exist)

ShadowNinja commented:
@s-l-teichmann That's sort of what the old system did (although it was multiplication-based).  The problem with that system was that it was very hard to do things like `DELETE FROM blocks WHERE abs(x) > 3 OR abs(y) > 3 OR abs(z) > 3`, and if you managed to do that it would be very ugly and likely inefficient.

@Rogier-5 Wow, quite a bit slower actually.  Maybe it's checking the `UNIQUE` constraint or something like that.

s-l-teichmann commented:
@ShadowNinja This particular case can be done pretty efficient but it needs some code. The main question is: Is this corner case so important to make the main cases (fetching blocks from th db)  slower?

This  patch only makes queries/updates/deletes more convinient at SQL level but not necessarily improves the performance. Having separate columns + the unique index instead introduces a lot of extra bookkeeping to the database. Furthermore the spatial selectivness of this index is not really existing.

The old code simply concats the coordinates and does not interleave them which makes a big difference.

Rogier-5 commented:
@ShadowNinja I did some additional research, and I second s-l-teichmann in that a lot of extra bookkeeping is introduced by the (unavoidable) extra index, resulting in extra disk IO as well...

From https://www.sqlite.org/lang_createtable.html#rowid, I get that for sqlite, the most efficient option is to have an INTEGER primary key (not INT, as it is now). I did some tests, and changing INT to INTEGER yields a small migration performance improvement:

> INTEGER: sqlite -> leveldb: 303.95user 162.98system 10:12.78elapsed 76%CPU
> INTEGER: leveldb -> sqlite: 267.00user 36.28system 6:17.50elapsed 80%CPU

ShadowNinja commented:
I'll try to split out the related fixes from this pull and add them separately, then work on optimizing the schema.

ShadowNinja commented:
Fixes merged.  I'll try to rebase this...

ShadowNinja commented:
Rebased, however **I'm seeing some weird corruption coming from the migrator** that I'm used to seeing with a broken getInetgerAsBlock implementation (eg, x < 0 is shifted up by one).

@PilzAdam: This now uses a compound primary key.

@Rogier-5: Migration is about 86x faster now, so actual backend speed differences should show up better now.  Note that although Minetest itself might get worse performace, other applications like mappers may be a lot faster (eg, one 40-line Python scipt that I wrote to clear blocks far from the origin that looped over every block in the database could be done with one SQL query, and it would be faster by orders of magnatude).  You should notice a speed improvement with `minetestmapper.py` (unless the DB time is insignificant compared to the overall run time in both cases).

PilzAdam commented:
@ShadowNinja :+1:; although you removed the section about "The Key", but it's still useful for the other backends (as you stated a few lines above).

ShadowNinja commented:
@PilzAdam At present only the SQLite3 backend is explained at all.  And the key documentation wasn't acurate for the LevelDB or Redis backends because they additionally encode it in an ASCII string.

Something similar to this PR should be done for the other backends too though, such as encoding the pos as a binary 6-character string for LevelDB.  I'm not sure if Redis can handle binary keys though.

Rogier-5 commented:
@ShadowNinja I have pulled and compiled the new version. I ran the database migration tests again, but I found two issues, which prevent some of the tests from executing:

* filesys.cpp should be compiled with _FILE_OFFSET_BITS defined to 64 (before the Posix headers are included). I don't know about windows or 64-bit linux, but on 32-bit linux, the stat() call in PathExists() always fails for files larger than 2G if _FILE_OFFSET_BITS is not defined to 64. That means that openDatabase will fail to detect that the database already exists, and attempt to create it - which fails obviously.

* On my system, checkMigrate fails when migrating from leveldb to sqlite3 (assertion failed when executing ALTER TABLE); my impression is that the code tries to migrate the sqlite3 database, which does not yet exist at that point...

With regard to the use of a configuration file setting to determine which database table format to use:
would it be an option to just use a different table name for the migrated database (e.g. blocks_xyz or so) instead of the configuation file entry ? Then the migration status could be detected by checking the names of the tables: if 'blocks' exists and 'blocks_xyz' doesn't, then migration hasn't started. If 'blocks' does not exist and 'blocks_xyz' does, then migration has completed, etc. This is more robust as the *actual* state of the database is used, instead of a presumed state. This might even allow one format or the other to be used, so that server operators can choose the desired table format for their world, *and* choose whether and when to migrate the database (migrating a large world database is a lengthy process, and when I tried to migrate a 10GB world, the database ballooned to 14G, filling up my disk before migration could complete).

As an aside: my choice for minetestmapper will be to detect the actual table format, and ignore the configuration file setting altogether. The SQL detection code for table names could be something like: `SELECT name FROM sqlite_master WHERE type = 'table' AND name NOT LIKE 'sqlite_%'`
Alternatively, using the current implementation, I obtain a list of table columns using: `PRAGMA table_info(blocks)`, and use that to determine the actual table format.

Rogier-5 commented:
@ShadowNinja In-place conversion speed improved 25% to 16 minutes:

> 299.94user 226.48system 16:32.01elapsed 53%CPU

As mentioned, leveldb <--> sqlite3 conversion failed.

Rogier-5 commented:
I ~~hacked~~patched the code so that the tests would run, and did a retest, migrating the same world as previously.

To make these results more (but not fully) comparable to my previous tests, I modified main.cpp so that database saves are performed every 10 seconds, instead of every second.

> Original: sqlite3 -> leveldb: 304.92user 155.85system 11:27.11elapsed 67%CPU
> Original: leveldb -> sqlite3: 273.94user 43.55system 6:33.43elapsed 80%CPU

> New: sqlite3 -> leveldb: 743.24user 307.06system 31:16.71elapsed 55%CPU
> New: leveldb -> sqlite3: 298.42user 162.30system 12:33.44elapsed 61%CPU

As my previous tests were already performed with an additional patch to minetestserver for greater database migration speed, I wasn't able to obtain an improvement, relative to my previous tests, similar to the 86x that you got.

For comparison: the same results with endSave/beginSave being done every second.

> Original: sqlite3 -> leveldb: 302.56user 157.94system 11:58.32elapsed 64%CPU
> Original: leveldb -> sqlite3: 271.88user 44.86system 7:06.28elapsed 74%CPU

> New: sqlite3 -> leveldb: 732.92user 304.34system 30:30.78elapsed 56%CPU
> New: leveldb -> sqlite3: 304.97user 202.70system 19:02.29elapsed 44%CPU

ShadowNinja commented:
@Rogier-5: Are you sure your build includes e9eda2b0d0019890cd404e4af25b7adf349e288f?  It should be massively faster, but your times seem almost the same as before.

Reading the schema from the DB seems rather ugly to me.  There are a lot of variables it check instead of just one integer.

I'll fix the issue you mentioned soon...

Rogier-5 commented:
@ShadowNinja I regret I must confirm that my build *does* include e9eda2b... For my first tests already, I had tweaked the server to improve the migration speed: by significantly reducing the frequency of commits and avoiding block (de)serialisation. Without those changes, the first tests would have lasted hours - *each*. I do have an SSD, but for the rest, my computer is not that fast... So your patch e9eda2b did improve upon my changes, but not by orders of magnitude...

>Reading the schema from the DB seems rather ugly to me. There are a lot of variables it check instead of just one integer.

If both tables have different names, it's just one query, and reading one row containing one string  (or two rows if migration is in progress), and checking if its value is 'blocks' or 'blocks_xyz' (or whatever the name). ([Example code](https://gist.github.com/Rogier-5/4f1aa7f5b36055ddc5ac)). Seems quite neat to me, but I guess that's a matter of personal taste.

If the tables have the same name, the query `PRAGMA table_info('blocks')` returns one row for every column. Simply counting the number of rows (i.e. columns in the table) would be sufficient. Additionally, a second query for `blocks_old` could determine whether a migration was in progress.

While reading through the sqlite docs, I also found an option which *is* one integer, and which might be preferable. It is also more robust than a configuration file setting, which may become mangled, and it's even specifically designed for this purpose: [`PRAGMA user_version`](http://www.sqlite.org/sessions/pragma.html#pragma_schema_version). This is basically the same as a variable in `world.mt`, except that it's stored in the database itself, so it can't get mangled if something happens to the `world.mt` file.

A last alternative, which is maybe the most simple and straightforward possible, and IMHO also the cleanest, is to just `prepare()` a query for each of the possible table formats. Depending on which `sqlite_prepare()` fails, you know whether the database has been migrated or not, or even whether a migration is in progress.

orwell96 commented:
Can we pick this up again?
It would make things much easier, and would be the first step to larger worlds.
(if we should ever need this)

ShadowNinja commented:
Updated and rebased.  Now uses `PRAGMA user_version` instead of `world.mt`.  We may want to go even farther and use SQLite's R*Tree implementation.  That's likely to offer a decent speed improvement for regular fetching operations (although insert will likely be slower).

ShadowNinja commented:
Update: Added R\*Tree support.  It's slower for me, but I'm testing on a tiny database and the strength of R\*Trees is their scalability, so we'll have to test with a very large database.

sfan5 commented:
:+1: for concept

celeron55 commented:
Concept is good but I don't think we can accept that much worse performance for minetestmapper.

ShadowNinja commented:
Rebased.  I'll take a look at performance soon<sup>tm</sup>.

Fixer-007 commented:
This will help so much to deal with world database, like parsing it, unexploring, crashing, mapping...

rubenwardy commented:
From reading the thread, it looks like it's slower for everyday use but faster and easier for editing externally/running the mapper. Is this true?

It's likely not worth reducing performance just for external tools.

I'm guessing the convertor needs to generate block IDs where it didn't need to before. Could this be the cause of the slow down?

nerzhul commented:
Benefit is maintenance, and also if map size is increased it solves a Magic id calculation problem. A b-tree on 3 fields is a little bit slower than 1 but it's not the bottleneck in map loading, it's the compression

paramat commented:
> This also switches to RTree based storage. This may improve performance, but it will only be substantial with very big databases. Small databases will be slower with the RTr

> Update: Added R*Tree support. It's slower for me, but I'm testing on a tiny database and the strength of R*Trees is their scalability, so we'll have to test with a very large database.

Most worlds are small databases, and the small worlds are usually run on the slower devices. The larger worlds are on more powerful servers where performance isn't so critical.

> Concept is good but I don't think we can accept that much worse performance for minetestmapper. I suggest getting new results for minetestmapper and see if it performs well enough with the latest changes.

Please can we attend to this?

The listed advantages are nice tohave but i feel possibly not worth it if there is performance loss in small worlds, which there seems to be.

> and also if map size is increased

Don't consider that for this PR, it's unlikely to happen, if it does at all it could be years away.

@ShadowNinja or others please can you attend to this?

nerzhul commented:
Don't forget one thing, for multiplayer servers PostgreSQL is the most powerful backend

paramat commented:
@ShadowNinja does this need adopting, since you have little time?

paramat commented:
Anyone interested in adopting this? Closing in 1 month if no interest shown.

AntonBoch1244 commented:
@ShadowNinja your need to update conflicting files for merge, please do that, or close PR.

SmallJoker commented:
@AntonBoch1244 Adoption will happen somewhen, as the label already suggests. As this is a quite important PR/idea, this will be kept open until it is either rebased or replaced by another PR.

paramat commented:
I should clarify that the author has no time for MT currently so this is marked for adoption, therefore rebasing this is pointless.

rubenwardy commented:
No response from author, unclear benefits, and massively out of date

paramat commented:
I support a close. No adoption interest either.
